### DISCUSSION TITLE
How to Perform Searches with Expansion Using PocketBase in Go?

### DISCUSSION THREAD

**Discussion by uazapi** [pocketbase/pocketbase#5471] [Q&A] [answered]
Hello,

I'm using PocketBase in a project to create an API with Go. In my application, I need to perform searches on my collections and, at the same time, expand related fields in a single query, similar to the expand functionality available in the JavaScript SDK.

In the JavaScript SDK, I can easily achieve this with the expand method, but I haven't found a direct way to do this using Go. Currently, I'm performing searches on the main collection and then running separate queries to expand the related fields. However, this approach can be inefficient for large datasets.

My question is: Is there a native or recommended way to perform searches with related field expansion using PocketBase in Go? If not, what would be the most efficient approach to achieve this?

Thank you in advance for your help, and congratulations on the excellent project!

Best regards,

**Comment by ganigeorgiev** [comment, marked as answer]
The SDKs search functionality can be emulated using the [`FindRecordsByFilter` method](https://pocketbase.io/docs/go-records/#fetch-multiple-records).

In order to expand the relations of the fetched records you have several options depending on your execution context, but the most generic one would be to call `ExpandRecords` as shown in [Programmatically expanding relations](https://pocketbase.io/docs/go-records/#programmatically-expanding-relations):

```go
// NB! this doesn't check rules since we don't have any information about the request or the authenticated state
// if you want to restrict the expand you can specify an explicit expand function as last argument OR call apis.EnrichRecords
errs := app.Dao().ExpandRecords(records, []string{"rel1", "rel2.subrel"}, nil);
if len(errs) > 0 {
    return fmt.Errorf("failed to expand: %v", errs)
}
```

If you want to set a default expand as part of a request, then you can call the [Enrich records routing helpers](https://pocketbase.io/docs/go-routing/#enrich-records):
```go
apis.EnrichRecords(c, app.Dao(), records, "rel1", "rel2.sub")
```
